File: easy/valid-parentheses.ts
// The first submission -> 97ms, 44.7MB
/*
function isValid(s: string): boolean {
    const temp = []
    let t
    for (const c of s) {
        if (c == "(" || c == "[" || c == "{") {
            temp.push(c)
            continue
        }
        t = temp.pop()
        if (c == ")" && t != "(") {
            return false
        }
        if (c == "]" && t != "[") {
            return false
        }
        if (c == "}" && t != "{") {
            return false
        }
    }
    return temp.length == 0
};
*/

// The second submission -> 74ms, 44.3MB
const closerParenthesesMap = {
    ")": "(",
    "]": "[",
    "}": "{",
}

const opernerParentheses = ["(", "[", "{"];

function isValid(s: string): boolean {
    const temp = []
    for (const c of s) {
        if (opernerParentheses.includes(c)) {
            temp.push(c)
            continue
        }
        if (temp.pop() != closerParenthesesMap[c]) {
            return false
        }
    }
    return temp.length == 0
};

File: easy/valid-parentheses.java
// The third submission -> 2ms, 42.2mb
class Solution {
    public boolean isValid(String s) {
        char[] chars = s.toCharArray();
        int len = s.length();

        Stack<Character> temp = new Stack<Character>();
        for (int i = 0; i < len; i++) {
            char c = chars[i];

            if (c == '(' || c == '[' || c == '{') {
                temp.push(c);
                continue;
            }

            if (temp.empty()) {
                return false;
            }

            char last = temp.pop();

            if (c == ')' && last != '(') {
                return false;
            }
            if (c == ']' && last != '[') {
                return false;
            }
            if (c == '}' && last != '{') {
                return false;
            }
        }

        return temp.empty();
    }
}